### DISCUSSION TITLE
impl Responder for String conflict

### DISCUSSION THREAD

**Discussion by SardineFish** [actix/actix-web#1952] [Q&A] [answered]
I want to implement `Responder` for some type to customize the handler output, for example
```rust
#[get("/foo")]
fn foo() -> u32 { }
#[get("/bar")]
fn bar() -> String { }
#[get("/baz")]
fn baz() -> Option<MyData, MyError> { }

impl Responder for u32 {} // ...
impl Responder for String {} // ... 
impl Responder for Option<MyData, MyError> {} // ...
```
But some was conflict with the implement in crate `actix_web` such as `String`.

Is there other or better way to implement some thing like post processing for data returned by handler?

**Comment by fakeshadow** [comment, marked as answer]
wrap your string in a new type and impl responder for it.

**Reply by SardineFish** [reply]
Sure, it is a nice solusion, but not elegant I think.

**Reply by robjtede** [reply]
Newtype pattern is common and reasonably elegant solution to overlapping impls.

If you look at the docs for the Responder trait it lets you customize the response code and headers with extensions methods.

**Reply by SardineFish** [reply]
Thanks, I got it.
I acturally want to write handler like `fn foo() -> u32`  instead of `fn foo() -> MyResponder` to make it more readable. And I found it impossible due to rust compiller restriction. I'm now trying to achive this by writing a proc macro to wrap my custom responder : )

**Reply by robjtede** [reply]
`fn foo() -> impl Responder`  
`fn foo() -> actix_web::Result<impl Responder>`

is the most flexible way to declare handlers

**Reply by SardineFish** [reply]
I finally implement my custom responder and writing my handler like following. It looks complicate, and can be simplify by proc macro. But that need to call the proc macro `actix_web::get` or `actix_web_codegen::route`, and it was not `pub`. Will it be exposed publicly to let others customize router macro?
 
```rust
#[actix_web::get("/")]
async fn get_foo() -> MyResponder<u32> {
    run(async move {
        // Actural controller codes...
        let n = async_func_with_result_u32().await?;
        Ok(n)
    }).await
}

impl<T: Serialize> Responder for MyResponder<T> {
    // ...
}

pub async fn run<R : Serialize, F: Future<Output = Result<R, Error>>>(async_block: F) -> MyResponder<R> {
    let result = async_block.await;
    MyResponder::<R>::from(result)
}

```

**Reply by robjtede** [reply]
I don't particularly understand what you're trying to achieve.

**Reply by SardineFish** [reply]
process `Result<T>` returned from my handler into `HttpResponse`

**Reply by robjtede** [reply]
`Result<T: Responder, impl Into<actix_web::Error>>` is already a Responder

**Reply by SardineFish** [reply]
Yes, but it will convert all `Err(Error)` into `actix_web::Error`, which is not I expect to.
